models/pagination.go

```diff
@@ -1,35 +1,142 @@
 package models
 
-import "strconv"
+import (
+	"math"
+	"slices"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+const defaultOffset = 0
+const defaultLimit = 20
+const defaultSort = "default"
 
 type Filter struct {
 	Name, Value string
 }
 
 type Pagination struct {
-	limit   int
 	offset  int
+	limit   int
 	sort    string
+	total   int
 	filters []Filter
 }
 
-func NewPagination(sort string, filters ...Filter) *Pagination {
-	if sort == "default" {
-		sort = ""
+func NewPagination(offset int, limit int, sort string, filters ...Filter) *Pagination {
+	if offset < 0 {
+		offset = defaultOffset
+	}
+
+	if limit <= 0 {
+		limit = defaultLimit
 	}
 
 	return &Pagination{
+		offset:  offset,
+		limit:   limit,
+		total:   -1,
 		sort:    sort,
 		filters: filters,
 	}
 }
 
+func (p *Pagination) Offset() int {
+	return p.offset
+}
+
 func (p *Pagination) Limit() int {
 	return p.limit
 }
 
-func (p *Pagination) Offset() int {
-	return p.offset
+func (p *Pagination) Total() int {
+	return p.total
+}
+
+func (p *Pagination) SetTotal(total int) {
+	p.total = total
+}
+
+func (p *Pagination) StartOfPage() int {
+	return p.Offset() + 1
+}
+
+func (p *Pagination) EndOfPage() int {
+	if p.Total() >= 0 {
+		return min(p.Total(), p.Offset()+p.Limit())
+	}
+
+	return p.Offset() + p.Limit()
+}
+
+func (p *Pagination) CurrentPage() int {
+	if p.Total() < 0 {
+		return -1
+	}
+
+	return int(math.Floor(float64(p.Offset())/float64(p.Limit()))) + 1
+}
+
+func (p *Pagination) NumberOfPages() int {
+	if p.Total() < 0 {
+		return -1
+	}
+
+	return int(math.Ceil(float64(p.Total()) / float64(p.Limit())))
+}
+
+func (p *Pagination) PageOffset(page int) int {
+	return (page - 1) * p.Limit()
+}
+
+func (p *Pagination) VisiblePages() []int {
+	if p.total < 0 {
+		return []int{1}
+	}
+
+	// Initialize map with each possible page number
+	pages := make(map[int]bool, p.NumberOfPages())
```

Comment:
small detail: better to use a map[int]struct{} here, the empty struct takes no memory